example of how compilation failed:
```
~/github/romani/checkstyle [mdhamed238/issue-20877-resolve-compilation-errors-in-inputs L|✚ 1] 
$ git diff
diff --git a/src/it/resources/org/checkstyle/suppressionxpathfilter/blocks/needbraces/InputXpathNeedBracesSingleLine.java b/src/it/resources/org/checkstyle/suppressionxpathfilter/blocks/needbraces/InputXpathNeedBracesSingleLine.java
index 67abc90f49..3606270760 100644
--- a/src/it/resources/org/checkstyle/suppressionxpathfilter/blocks/needbraces/InputXpathNeedBracesSingleLine.java
+++ b/src/it/resources/org/checkstyle/suppressionxpathfilter/blocks/needbraces/InputXpathNeedBracesSingleLine.java
@@ -4,7 +4,7 @@ import com.puppycrawl.tools.checkstyle.checks.blocks.needbraces.InputNeedBracesS
 
 public class InputXpathNeedBracesSingleLine {
     private static class SomeClass {
-        boolean flag = true;
+        boolean flag true;
         private static boolean test(boolean k) {
             return k;
         }

✔ ~/github/romani/checkstyle [mdhamed238/issue-20877-resolve-compilation-errors-in-inputs L|✚ 1] 
$ ./mvnw clean compile test-compile
[INFO] Scanning for projects...
[INFO] Inspecting build with total of 1 modules
[INFO] Installing Central Publishing features
[INFO] 
[INFO] ------------------< com.puppycrawl.tools:checkstyle >-------------------
[INFO] Building checkstyle 13.9.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.5.0:clean (default-clean) @ checkstyle ---
[INFO] Deleting /home/roman/github/romani/checkstyle/target
[INFO] 
[INFO] --- tidy:1.4.0:check (validate) @ checkstyle ---
[INFO] 
[INFO] --- enforcer:3.6.3:enforce (enforce-maven) @ checkstyle ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] 
[INFO] --- enforcer:3.6.3:enforce (enforce-versions) @ checkstyle ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] Rule 1: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] Rule 2: org.apache.maven.enforcer.rules.dependency.DependencyConvergence passed
[INFO] 
[INFO] --- xml:1.2.1:check-format (default) @ checkstyle ---
[INFO] 
[INFO] --- antlr4:4.13.2:antlr4 (default) @ checkstyle ---
[INFO] ANTLR 4: Processing source directory /home/roman/github/romani/checkstyle/src/main/resources
[INFO] Processing grammar: com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4
[INFO] Processing grammar: com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4
[INFO] Processing grammar: com/puppycrawl/tools/checkstyle/grammar/javadoc/JavadocCommentsLexer.g4
[INFO] Processing grammar: com/puppycrawl/tools/checkstyle/grammar/javadoc/JavadocCommentsParser.g4
[INFO] 
[INFO] --- build-helper:3.6.1:add-source (add-source) @ checkstyle ---
[INFO] Source directory: /home/roman/github/romani/checkstyle/target/generated-sources/antlr added.
[INFO] 
[INFO] --- resources:3.5.0:resources (default-resources) @ checkstyle ---
[INFO] Copying 417 resources from src/main/resources to target/classes
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-test-source) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/test/resources added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-it-test-source) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/it/java added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-it-test-resource) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/it/resources added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-xdocs-examples-source) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/xdocs-examples/java added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-xdocs-examples-resource) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/xdocs-examples/resources added.
[INFO] 
[INFO] --- compiler:3.15.0:compile (default-compile) @ checkstyle ---
[INFO] Recompiling the module because of changed source code.
[INFO] Compiling 487 source files with javac [debug release 21] to target/classes
[INFO] /home/roman/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java: Some input files use or override a deprecated API.
[INFO] /home/roman/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java: Recompile with -Xlint:deprecation for details.
[INFO] /home/roman/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java: /home/roman/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java uses unchecked or unsafe operations.
[INFO] /home/roman/github/romani/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/AbstractAutomaticBean.java: Recompile with -Xlint:unchecked for details.
[INFO] 
[INFO] --- tidy:1.4.0:check (validate) @ checkstyle ---
[INFO] 
[INFO] --- enforcer:3.6.3:enforce (enforce-maven) @ checkstyle ---
[INFO] 
[INFO] --- enforcer:3.6.3:enforce (enforce-versions) @ checkstyle ---
[INFO] Rule 2: org.apache.maven.enforcer.rules.dependency.DependencyConvergence passed
[INFO] 
[INFO] --- xml:1.2.1:check-format (default) @ checkstyle ---
[INFO] 
[INFO] --- antlr4:4.13.2:antlr4 (default) @ checkstyle ---
[INFO] No grammars to process
[INFO] ANTLR 4: Processing source directory /home/roman/github/romani/checkstyle/src/main/resources
[INFO] 
[INFO] --- build-helper:3.6.1:add-source (add-source) @ checkstyle ---
[INFO] Source directory: /home/roman/github/romani/checkstyle/target/generated-sources/antlr added.
[INFO] 
[INFO] --- resources:3.5.0:resources (default-resources) @ checkstyle ---
[INFO] Copying 417 resources from src/main/resources to target/classes
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-test-source) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/test/resources added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-it-test-source) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/it/java added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-it-test-resource) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/it/resources added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-xdocs-examples-source) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/xdocs-examples/java added.
[INFO] 
[INFO] --- build-helper:3.6.1:add-test-source (add-xdocs-examples-resource) @ checkstyle ---
[INFO] Test Source directory: /home/roman/github/romani/checkstyle/src/xdocs-examples/resources added.
[INFO] 
[INFO] --- compiler:3.15.0:compile (default-compile) @ checkstyle ---
[INFO] Nothing to compile - all classes are up to date.
[INFO] 
[INFO] --- plexus-component-metadata:2.2.0:generate-metadata (default) @ checkstyle ---
[INFO] Discovered 10 component descriptor(s)
[INFO] 
[INFO] --- exec:3.6.3:java (default) @ checkstyle ---
[INFO] 
[INFO] --- exec:3.6.3:java (generate-search-index) @ checkstyle ---
[INFO] 
[INFO] --- resources:3.5.0:testResources (default-testResources) @ checkstyle ---
[INFO] Copying 3636 resources from src/test/resources to target/test-classes
[INFO] 
[INFO] --- compiler:3.15.0:testCompile (default-testCompile) @ checkstyle ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 5986 source files with javac [debug release 21] to target/test-classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/roman/github/romani/checkstyle/src/it/resources/org/checkstyle/suppressionxpathfilter/blocks/needbraces/InputXpathNeedBracesSingleLine.java:[7,21] ';' expected
[INFO] 1 error
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
```

--- a/.ci/release-maven-perform.sh
+++ b/.ci/release-maven-perform.sh
@@ -15,7 +15,7 @@ checkForVariable "GITHUB_REPOSITORY_OWNER"
 TARGET_VERSION=$1
 echo TARGET_VERSION="$TARGET_VERSION"
 
-SKIP_TEST="-DskipTests -DskipITs"
+SKIP_TEST="-Dcheckstyle.skipCompileInputResources=true -DskipTests -DskipITs"
 SKIP_CHECKSTYLE="-Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true"
 SKIP_OTHERS="-Dpmd.skip=true -Dspotbugs.skip=true -Djacoco.skip=true -Dxml.skip=true \
   --no-transfer-progress"

--- a/.ci/release-maven-prepare.sh
+++ b/.ci/release-maven-prepare.sh
@@ -28,7 +28,7 @@ if [[ $(grep "<section name=\"Release $TARGET_VERSION\">" src/site/xdoc/releasen
   exit 1
 fi
 
-SKIP_TEST="-DskipTests -DskipITs"
+SKIP_TEST="-Dcheckstyle.skipCompileInputResources=true -DskipTests -DskipITs"
 SKIP_CHECKSTYLE="-Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true"
 SKIP_OTHERS="-Dpmd.skip=true -Dspotbugs.skip=true -Djacoco.skip=true -Dxml.skip=true -Dgpg.skip"
 

--- a/.ci/run-link-check-plugin.sh
+++ b/.ci/run-link-check-plugin.sh
@@ -8,6 +8,7 @@ uname -a
 ./mvnw --version
 curl --fail-with-body -I https://sourceforge.net/projects/checkstyle/
 ./mvnw -e --no-transfer-progress clean site -Dcheckstyle.ant.skip=true -DskipTests -DskipITs \
+   -Dcheckstyle.skipCompileInputResources=true \
    -Dpmd.skip=true -Dspotbugs.skip=true -Djacoco.skip=true -Dcheckstyle.skip=true
 mkdir -p .ci-temp
 

--- a/.ci/validation.cmd
+++ b/.ci/validation.cmd
@@ -21,7 +21,9 @@ if "%OPTION%" == "run_checkstyle" (
 )
 
 if "%OPTION%" ==  "verify_without_checkstyle" (
-  call mvnw.cmd -e --no-transfer-progress verify -Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true^
+  :: Appveyor has old jdk21 there is problem compilation of Inputs
+  call mvnw.cmd -e --no-transfer-progress verify -Dcheckstyle.skipCompileInputResources=true^
+    -Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true^
     || goto :ERROR
   goto :END_CASE
 )

--- a/.ci/validation.sh
+++ b/.ci/validation.sh
@@ -613,6 +613,7 @@ release-dry-run)
     ./mvnw -e --no-transfer-progress release:prepare -DdryRun=true --batch-mode \
     -Darguments='-DskipTests -DskipITs -Djacoco.skip=true -Dpmd.skip=true \
       -Dspotbugs.skip=true -Dxml.skip=true -Dcheckstyle.ant.skip=true \
+      -Dcheckstyle.skipCompileInputResources=true \
       -Dcheckstyle.skip=true -Dgpg.skip=true --no-transfer-progress'
     ./mvnw -e --no-transfer-progress release:clean
   fi
@@ -714,13 +715,6 @@ check-since-version)
   else
     echo "No new Check, all is good."
   fi
-  ;;
-
-compile-test-resources)
-  # this task is useful during migration to new JDK to let compile resources on new jdk only
-  ./mvnw -e --no-transfer-progress clean test-compile \
-  -Dcheckstyle.skipCompileInputResources=false \
-  -Dmaven.compiler.release=21
   ;;
 
 javac21-exceptional)

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -323,10 +323,6 @@ workflows:
           image-name: cimg/openjdk:21.0.6
           command: ./.ci/validation.sh javac21-exceptional
       - validate-with-script:
-          name: java 21 test resources compile
-          image-name: cimg/openjdk:21.0.8
-          command: ./.ci/validation.sh compile-test-resources
-      - validate-with-script:
           name: javac25
           image-name: cimg/openjdk:25.0
           command: ./.ci/validation.sh javac25

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,4 +1,4 @@
-image: Visual Studio 2019
+image: Visual Studio 2026
 version: '{build}'
 skip_tags: true
 

--- a/pom.xml
+++ b/pom.xml
@@ -217,7 +217,7 @@
     </sevntu-checkstyle-check.checkstyle.version>
     <versions.plugin.version>2.21.0</versions.plugin.version>
     <compiler.plugin.version>3.15.0</compiler.plugin.version>
-    <checkstyle.skipCompileInputResources>true</checkstyle.skipCompileInputResources>
+    <checkstyle.skipCompileInputResources>false</checkstyle.skipCompileInputResources>
     <checkstyle.openrewrite.version>1.0.0-SNAPSHOT</checkstyle.openrewrite.version>
     <java.version>21</java.version>
     <maven.compiler.release>${java.version}</maven.compiler.release>
@@ -2173,6 +2173,16 @@
               <excludes>
                 <exclude>**/Example*</exclude>
                 <exclude>**/Input*</exclude>
+                <!-- input helper classes co-located in Input*/Example* files (or named
+                     UseCase*) that do not match the Input*/Example* naming; compiled since
+                     checkstyle.skipCompileInputResources=false and use forbidden apis by
+                     design as test fixtures -->
+                <exclude>**/UseCase*</exclude>
+                <exclude>**/classdataabstractioncoupling/MyRecord*</exclude>
+                <exclude>**/whitespacearound/foo3*</exclude>
+                <exclude>**/missingoverrideonrecordaccessor/Person*</exclude>
+                <exclude>**/missingoverrideonrecordaccessor/Document*</exclude>
+                <exclude>**/rule61overridealwaysused/Document*</exclude>
                 <!-- usage of system output by design -->
                 <exclude>**/MainTest.class</exclude>
                 <!-- tests need forbidden apis to test the main code fully -->
@@ -2607,6 +2617,7 @@
       <id>no-validations</id>
       <properties>
         <skipTests>true</skipTests>
+        <checkstyle.skipCompileInputResources>true</checkstyle.skipCompileInputResources>
         <checkstyle.ant.skip>true</checkstyle.ant.skip>
         <pmd.skip>true</pmd.skip>
         <spotbugs.skip>true</spotbugs.skip>

